opencensus/context/with_context.h

```diff
@@ -0,0 +1,52 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef OPENCENSUS_CONTEXT_WITH_CONTEXT_H_
+#define OPENCENSUS_CONTEXT_WITH_CONTEXT_H_
+
+#include "opencensus/context/context.h"
+
+namespace opencensus {
+namespace context {
+
+// WithContext is an RAII object, only ever stack-allocate it. While it's in
+// scope, the execution will happen under a copy of the given Context.
+class WithContext {
+ public:
+  explicit WithContext(const Context& ctx);
+  explicit WithContext(Context&& ctx);
+  ~WithContext();
+
+ private:
+  WithContext() = delete;
+  WithContext(const WithContext&) = delete;
+  WithContext(WithContext&&) = delete;
+  WithContext& operator=(const WithContext&) = delete;
+  WithContext& operator=(WithContext&&) = delete;
+
+  Context swapped_context_;
```

Comment:
Based on some experience that I have from dealing with other context implementation it is better if we have a way to check that objects are allocated and deallocated on the same thread. I suggest to put this check under the NDEBUG macro.

Reply:
Filed https://github.com/census-instrumentation/opencensus-cpp/issues/203